### ISSUE TITLE
C/C++ incorrect syntax region

### ISSUE THREAD

**Issue by idnsunset** [marshallward/vim-restructuredtext#55]
Hi,
I attempted to add `'c': ['c']` to the `g:rst_syntax_code_list` to enable C syntax highlighting but run into the issue with c and cpp code blocks as described below.

Given the following rst text:

```rst
.. code:: cpp
   using namespace std;

   int main() {}

using
```

Both 'using' are highlighted as C++ statement.

**Comment by marshallward** [comment]
Yes, this is something that I had noticed as well, but looks like I never logged the issue.  Thanks for submitting it.

When I tried to sort this out, it seemed to be a problem with one of those syntax files, but I could not make any more progress at the time.

Your example is a bit odd in that there should be a blank line between the directive and the code block.  That may be affecting things as well here.

**Comment by madjxatw** [comment]
I've figured out what exactly happens there.

If you take a look into the C++ syntax file (`syntax/cpp.vim`), you will find that it loads `syntax/c.vim` by exectuing `runtime! syntax/c.vim`. When you put both 'c' and 'cpp' in `g:rst_syntax_code_list`, the `c.vim` syntax file will be sourced twice, resulting in the C syntax group having duplicate patterns.

Taking the example you posted, place the cursor at the opening `{` and run the following command:

```vim
:echo map(synstack(line('.'), col('.')), 'synIDattr(v:val, "name")')
```

and you will get:

```vim
['rstDirectivecpp', 'cBlock', 'cBlock']
```

This means that the `cBlock` syntax group is matched twice at `{`.

Now lets check the definition of `cBlock` group by running:

```vim
:syn list @rstcpp
```

and you will get:

    --- Syntax items ---
    cBlock         xxx start=/{/ end=/}/  contained transparent fold 
                       start=/{/ end=/}/  contained transparent fold

You can see the duplicate patterns defined for `cBlock`.

When reaching the closing `}`, the inner `cBlock` group ends but the remaining texts are still in the outer `cBlock` group until reaching another `}`, this is why the second `using` is highlighted.

I've thought about why 'c' is not added to `g:rst_syntax_code_list` and this might be the reason, i.e. C++ syntax file covers all C syntax by sourcing C syntax file. However, we don't want to get some C++ keywords such as `using` highlighted in a C code block, hence we will have to define a separate @rstc syntax cluster, meanwhile avoid source `syntax/c.vim` twice.

I attempted to fix this by introducing extra code specific for C when processing CPP syntax groups: https://github.com/madjxatw/vim-restructuredtext/blob/bbd78b20f39264c37673a12c30c9104e79ae9528/syntax/rst.vim#L255-L274

This might not be the best approach, welcome better solution.

**Comment by marshallward** [comment]
Thank you @madjxatw for solving this mystery, your explanation makes a lot of sense.  This looks like an interesting solution to a very difficult problem.

It would be nice if there were a way to catch the more general issue of recursive highlighting, since quite a few syntax languages do this.  For example, I think php inherits html!

In the meantime, I'll try to find some time to work through your solution.
